[BUG] document_renamer and document_sanity_checker report missing originals after storage path change

### Description

I imported 8 PDF files. Right after the import I changed them from "inbox" label to some other label. One label changes the storage_path value in a workflow. This works without any problems normally.

After the processing was done I clicked "preview" of the files and 2 of them is broken now. The files are moved correctly to the new destination like all others but the database still looks into the old destination.

### Steps to reproduce

1. Import multiple PDFs
2. Immediately click in the inbox and select a tag, that triggers another storage_path
3. Wait until paperless ui is done
4. click on the PDFs, some are broken here (maybe because of slow NAS)

### Webserver logs

```bash
[paperless-webserver-1] [2026-03-18 09:42:25,183] [WARNING] [django.request] Not Found: /api/documents/1304/preview/
[paperless-webserver-1] [2026-03-18 09:42:25,274] [WARNING] [django.request] Not Found: /api/documents/1304/preview/
[paperless-webserver-1] [2026-03-18 09:42:34,802] [WARNING] [django.request] Not Found: /api/documents/1304/preview/
[paperless-webserver-1] [2026-03-18 09:42:37,614] [WARNING] [django.request] Not Found: /api/documents/1298/preview/
[paperless-webserver-1] [2026-03-18 09:42:37,830] [WARNING] [django.request] Not Found: /api/documents/1298/preview/
[paperless-webserver-1] [2026-03-18 09:42:45,648] [WARNING] [django.request] Not Found: /api/documents/1298/preview/
[paperless-webserver-1] [2026-03-18 09:44:11,883] [WARNING] [django.request] Not Found: /api/documents/1298/preview/
```

Also document_renamer cannot fix it.

```
root@a012988ed3f7:/usr/src/paperless/src# document_renamer
 56% 545/976 [02:41<01:49,  3.93it/s][2026-03-18 09:58:42,929] [CRITICAL] [paperless.handlers] Document 2020-03-16 Beitragsbescheid_2006-00715_2020: File /usr/src/paperless/media/documents/originals/andre/2020/Beitragsbescheid_2006-00715_2020.pdf doesn't exist.
 62% 607/976 [02:59<01:52,  3.28it/s][2026-03-18 09:59:00,544] [CRITICAL] [paperless.handlers] Document 2019-03-15 Beitragsbescheid_2006-00715_2019: File /usr/src/paperless/media/documents/originals/andre/2019/Beitragsbescheid_2006-00715_2019.pdf doesn't exist.
```


```
root@a012988ed3f7:/usr/src/paperless/src# document_sanity_checker
100% 976/976 [00:08<00:00, 111.58it/s]
[2026-03-18 10:03:19,741] [INFO] [paperless.sanity_checker] Detected following issue(s) with document #1304, titled Beitragsbescheid_2006-00715_2020
[2026-03-18 10:03:19,746] [ERROR] [paperless.sanity_checker] Original of document does not exist.
[2026-03-18 10:03:19,749] [ERROR] [paperless.sanity_checker] Archived version of document does not exist.
[2026-03-18 10:03:19,756] [INFO] [paperless.sanity_checker] Detected following issue(s) with document #1298, titled Beitragsbescheid_2006-00715_2019
[2026-03-18 10:03:19,759] [ERROR] [paperless.sanity_checker] Original of document does not exist.
[2026-03-18 10:03:19,762] [ERROR] [paperless.sanity_checker] Archived version of document does not exist.
[2026-03-18 10:03:19,765] [WARNING] [paperless.sanity_checker] Orphaned file in media dir: /usr/src/paperless/media/documents/archive/gemeinsam/2020/Beitragsbescheid_2006-00715_2020.pdf
[2026-03-18 10:03:19,768] [WARNING] [paperless.sanity_checker] Orphaned file in media dir: /usr/src/paperless/media/documents/originals/gemeinsam/2019/Beitragsbescheid_2006-00715_2019.pdf
[2026-03-18 10:03:19,769] [WARNING] [paperless.sanity_checker] Orphaned file in media dir: /usr/src/paperless/media/documents/originals/gemeinsam/2020/Beitragsbescheid_2006-00715_2020.pdf
[2026-03-18 10:03:19,770] [WARNING] [paperless.sanity_checker] Orphaned file in media dir: /usr/src/paperless/media/documents/archive/gemeinsam/2019/Beitragsbescheid_2006-00715_2019.pdf
```

### Browser logs

```bash

```

### Services logs

```bash

```

### Paperless-ngx version

2.20.11

### Host OS

AlpineLinux with official Docker-Container

### Installation method

Docker - official image

### System status

```json
{
    "pngx_version": "2.20.11",
    "server_os": "Linux-6.18.16-0-lts-x86_64-with-glibc2.41",
    "install_type": "docker",
    "storage": {
        "total": 3836004532224,
        "available": 2006723854336
    },
    "database": {
        "type": "postgresql",
        "url": "paperless",
        "status": "OK",
        "error": null,
        "migration_status": {
            "latest_migration": "paperless_mail.0001_initial_squashed_0009_mailrule_assign_tags",
            "unapplied_migrations": []
        }
    },
    "tasks": {
        "redis_url": "redis://broker:6379",
        "redis_status": "OK",
        "redis_error": null,
        "celery_status": "OK",
        "celery_url": "celery@a012988ed3f7",
        "celery_error": null,
        "index_status": "OK",
        "index_last_modified": "2026-03-18T09:41:20.482379+01:00",
        "index_error": null,
        "classifier_status": "OK",
        "classifier_last_trained": "2026-03-18T09:05:02.257969Z",
        "classifier_error": null,
        "sanity_check_status": "ERROR",
        "sanity_check_last_run": "2026-03-18T09:11:55.320674Z",
        "sanity_check_error": "3 issues found. Check logs for details."
    },
    "websocket_connected": "OK"
}
```

### Browser

Firefox

### Configuration changes

_No response_

### Please confirm the following

- [x] I believe this issue is a bug that affects all users of Paperless-ngx, not something specific to my installation.
- [x] This issue is not about the OCR or archive creation of a specific file(s). Otherwise, please see above regarding OCR tools.
- [x] I have already searched for relevant existing issues and discussions before opening this report.
- [x] I have updated the title field above with a concise description.

--- a/src/documents/signals/handlers.py
+++ b/src/documents/signals/handlers.py
@@ -1,5 +1,6 @@
 from __future__ import annotations
 
+import hashlib
 import logging
 import shutil
 from pathlib import Path
@@ -391,6 +392,14 @@ class CannotMoveFilesException(Exception):
     pass
 
 
+def _path_matches_checksum(path: Path, checksum: str | None) -> bool:
+    if checksum is None or not path.is_file():
+        return False
+
+    with path.open("rb") as f:
+        return hashlib.md5(f.read()).hexdigest() == checksum
+
+
 def _filename_template_uses_custom_fields(doc: Document) -> bool:
     template = None
     if doc.storage_path is not None:
@@ -461,10 +470,12 @@ def update_filename_and_move_files(
             old_filename = instance.filename
             old_source_path = instance.source_path
             move_original = False
+            original_already_moved = False
 
             old_archive_filename = instance.archive_filename
             old_archive_path = instance.archive_path
             move_archive = False
+            archive_already_moved = False
 
             candidate_filename = generate_filename(instance)
             if len(str(candidate_filename)) > Document.MAX_STORED_FILENAME_LENGTH:
@@ -485,14 +496,23 @@ def update_filename_and_move_files(
                 candidate_source_path.exists()
                 and candidate_source_path != old_source_path
             ):
-                # Only fall back to unique search when there is an actual conflict
-                new_filename = generate_unique_filename(instance)
+                if not old_source_path.is_file() and _path_matches_checksum(
+                    candidate_source_path,
+                    instance.checksum,
+                ):
+                    new_filename = candidate_filename
+                    original_already_moved = True
+                else:
+                    # Only fall back to unique search when there is an actual conflict
+                    new_filename = generate_unique_filename(instance)
             else:
                 new_filename = candidate_filename
 
             # Need to convert to string to be able to save it to the db
             instance.filename = str(new_filename)
-            move_original = old_filename != instance.filename
+            move_original = (
+                old_filename != instance.filename and not original_already_moved
+            )
 
             if instance.has_archive_version:
                 archive_candidate = generate_filename(instance, archive_filename=True)
@@ -513,24 +533,38 @@ def update_filename_and_move_files(
                     archive_candidate_path.exists()
                     and archive_candidate_path != old_archive_path
                 ):
-                    new_archive_filename = generate_unique_filename(
-                        instance,
-                        archive_filename=True,
-                    )
+                    if not old_archive_path.is_file() and _path_matches_checksum(
+                        archive_candidate_path,
+                        instance.archive_checksum,
+                    ):
+                        new_archive_filename = archive_candidate
+                        archive_already_moved = True
+                    else:
+                        new_archive_filename = generate_unique_filename(
+                            instance,
+                            archive_filename=True,
+                        )
                 else:
                     new_archive_filename = archive_candidate
 
                 instance.archive_filename = str(new_archive_filename)
 
-                move_archive = old_archive_filename != instance.archive_filename
+                move_archive = (
+                    old_archive_filename != instance.archive_filename
+                    and not archive_already_moved
+                )
             else:
                 move_archive = False
 
             if not move_original and not move_archive:
-                # Just update modified. Also, don't save() here to prevent infinite recursion.
-                Document.objects.filter(pk=instance.pk).update(
-                    modified=timezone.now(),
-                )
+                updates = {"modified": timezone.now()}
+                if old_filename != instance.filename:
+                    updates["filename"] = instance.filename
+                if old_archive_filename != instance.archive_filename:
+                    updates["archive_filename"] = instance.archive_filename
+
+                # Don't save() here to prevent infinite recursion.
+                Document.objects.filter(pk=instance.pk).update(**updates)
                 return
 
             if move_original:

--- a/src/documents/tests/test_file_handling.py
+++ b/src/documents/tests/test_file_handling.py
@@ -1,4 +1,5 @@
 import datetime
+import hashlib
 import logging
 import tempfile
 from pathlib import Path
@@ -165,6 +166,52 @@ class TestFileHandling(DirectoriesMixin, FileSystemAssertsMixin, TestCase):
                 settings.ORIGINALS_DIR / "none" / "none.pdf",
             )
             self.assertEqual(document.filename, "none/none.pdf")
+
+    @override_settings(FILENAME_FORMAT=None)
+    def test_stale_save_recovers_already_moved_files(self) -> None:
+        old_storage_path = StoragePath.objects.create(
+            name="old-path",
+            path="old/{{title}}",
+        )
+        new_storage_path = StoragePath.objects.create(
+            name="new-path",
+            path="new/{{title}}",
+        )
+        original_bytes = b"original"
+        archive_bytes = b"archive"
+
+        doc = Document.objects.create(
+            title="document",
+            mime_type="application/pdf",
+            checksum=hashlib.md5(original_bytes).hexdigest(),
+            archive_checksum=hashlib.md5(archive_bytes).hexdigest(),
+            filename="old/document.pdf",
+            archive_filename="old/document.pdf",
+            storage_path=old_storage_path,
+        )
+        create_source_path_directory(doc.source_path)
+        doc.source_path.write_bytes(original_bytes)
+        create_source_path_directory(doc.archive_path)
+        doc.archive_path.write_bytes(archive_bytes)
+
+        stale_doc = Document.objects.get(pk=doc.pk)
+        fresh_doc = Document.objects.get(pk=doc.pk)
+        fresh_doc.storage_path = new_storage_path
+        fresh_doc.save()
+        doc.refresh_from_db()
+        self.assertEqual(doc.filename, "new/document.pdf")
+        self.assertEqual(doc.archive_filename, "new/document.pdf")
+
+        stale_doc.storage_path = new_storage_path
+        stale_doc.save()
+
+        doc.refresh_from_db()
+        self.assertEqual(doc.filename, "new/document.pdf")
+        self.assertEqual(doc.archive_filename, "new/document.pdf")
+        self.assertIsFile(doc.source_path)
+        self.assertIsFile(doc.archive_path)
+        self.assertIsNotFile(settings.ORIGINALS_DIR / "old" / "document.pdf")
+        self.assertIsNotFile(settings.ARCHIVE_DIR / "old" / "document.pdf")
 
     @override_settings(FILENAME_FORMAT="{correspondent}/{correspondent}")
     def test_document_delete(self):

